site/docs/v1/tech/apis/api-keys.adoc

```diff
@@ -0,0 +1,172 @@
+The unique Id of the API Key to retrieve.
+
+=== Response
+
+The response for this API contains a single API Key. The response is defined below along with an example JSON response.
+
+:never_search_error:
+include::docs/v1/tech/apis/_standard-get-response-codes.adoc[]
+:never_search_error!:
+
+include::docs/v1/tech/apis/_apikey-response-body.adoc[]
+
+
+== Create an API Key
+
+This API is used to create a new API Key. An API key with key manager permission set to true can create keys. A key with key manager permissions can not be created using this API. Only through admin UI or kickstart, it is possible to create such a key. An API key that is tenant scoped, can create another key for the same tenant.
+
+[.api-authentication]
+link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Create a new Key with the provided unique Id
+[.endpoint]
+.URI
+--
+[method]#POST# [uri]#/api/api-key/`\{apiKeyId\}`#
+--
+
+==== Request Parameters
+
+[.api]
+[field]#apiKeyId# [type]#[UUID]# [optional]#Optional# [default]#defaults to a generated UUID#::
+The unique Id of the API Key to create. If not specified a secure random UUID will be generated.
+
+include::docs/v1/tech/apis/_apikey-put-request-body.adoc[]
+
+=== Response
+
+The response for this API contains the Key that was updated.
+
+:never_search_error:
+include::docs/v1/tech/apis/_standard-post-response-codes.adoc[]
+:never_search_error!:
+
+include::docs/v1/tech/apis/_apikey-post-response-body.adoc[]
+
+
+== Create a New API Key From an Existing API Key
+
+This API is used to create a New API Key based on an existing API Key (source key). The API key that is being copied must not be set as a key manager, if it is, this API will return an error. The resulting new key will have the same permissions as the source key.
+
+[.api-authentication]
+link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Create a new Key from an Existing Key
+[.endpoint]
+.URI
+--
+[method]#PUT# [uri]#/api/api-key/#
+--
+
+.Example API Key request JSON
+[source,json]
+{
+  "sourceKeyId": "48b5762e-6829-4d7e-b53d-f7db5f5d3d1f"
+}
+
+
+=== Response
+
+The response for this API contains the Key that was created.
+
+:never_search_error:
+include::docs/v1/tech/apis/_standard-put-response-codes.adoc[]
+:never_search_error!:
+
+include::docs/v1/tech/apis/_apikey-response-body.adoc[]
+
+
+== Update an API Key
+
+This API is used to update an existing API Key. A tenant-scoped API key can update another API key for the same tenant.
+
+[.api-authentication]
+link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Update the Key with the given Id
+[.endpoint]
+.URI
+--
+[method]#PUT# [uri]#/api/api-key/`\{apiKeyId\}`#
+--
+
+=== Request Parameters
+
+[.api]
+[field]#keyId# [type]#[UUID]# [required]#Required#::
+The unique Id of the API Key to update.
+
+include::docs/v1/tech/apis/_apikey-put-request-body.adoc[]
```

Comment:
I also had to provide the "tenantId" when I was using a tenant scoped key master or I got an error.

Reply:
What error did you get? I am trying to reproduce this, but I do not get an error. I have a tenant scoped api key and I call PUT api without tenant id and it does update without modifying tenant id.

Reply:
Here's a shell script, running against released 1.26.1.

```
# tenant scoped key manager key. tenant is the `default` tenant. I had one other tenant in the instance. It's a super user key
API_KEY=soQdxM0_z8mU6lzWlT8yfEVuggb7obzBSHt9vowEGRfsT1oCLx4CbHoi

# another key scoped to the `default` tenant. Not a key manager.
OTHER_KEY=f4f15e04-818f-4b9a-b8ba-a8779952581e

curl -vvv -XPUT -H 'Content-type: application/json' -H "Authorization: $API_KEY" 'http://localhost:9011/api/api-key/'$OTHER_KEY -d '
{
  "apiKey": {
    "metaData": {
      "attributes": {
        "description": "new Key Description"
      }
    },
    "permissions": {
      "endpoints": {
        "/api/application": [
          "DELETE",
          "POST",
          "GET",
          "PUT",
          "PATCH"
        ]
      }
    }
  }
}
'
```

This gives me this error:

```
< ....
< HTTP/1.1 400 
< Cache-Control: no-store
< Content-Type: application/json;charset=UTF-8
< Content-Length: 170
< Date: Thu, 29 Apr 2021 23:08:32 GMT
< Connection: close
< 
* Closing connection 0
{"fieldErrors":{"tenantId":[{"code":"[required]tenantId","message":"The [tenantId] property is required when the API key you are using is limited to a single tenant."}]}}
```

When I add the `tenantId` field to the request body, it completes:

```
{"apiKey":{"id":"f4f15e04-818f-4b9a-b8ba-a8779952581e","insertInstant":1619631900103,"key":"06MtlBvDl8ZCOUylsPCujuCmHj6rg22zGejhSL5q3uuYCsyPPSoImHiJ","keyManager":false,"lastUpdateInstant":1619737910150,"metaData":{"attributes":{"description":"new Key Description"}},"permissions":{"endpoints":{"/api/application":["DELETE","POST","GET","PUT","PATCH"]}},"tenantId":"30663132-6464-6665-3032-326466613934"}}
```

Here's the updated script which works:

```
API_KEY=soQdxM0_z8mU6lzWlT8yfEVuggb7obzBSHt9vowEGRfsT1oCLx4CbHoi

OTHER_KEY=f4f15e04-818f-4b9a-b8ba-a8779952581e
curl -vvv -XPUT -H 'Content-type: application/json' -H "Authorization: $API_KEY" 'http://localhost:9011/api/api-key/'$OTHER_KEY -d '
{
  "apiKey": {
    "metaData": {
      "attributes": {
        "description": "new Key Description"
      }
    },
    "tenantId": "30663132-6464-6665-3032-326466613934",
    "permissions": {
      "endpoints": {
        "/api/application": [
          "DELETE",
          "POST",
          "GET",
          "PUT",
          "PATCH"
        ]
      }
    }
  }
}
'
```

HTH

Reply:
See this comment by Daniel. This is my understanding as well. 

https://github.com/FusionAuth/fusionauth-site/pull/593#discussion_r619525671

Reply:
Just had a quick slack call with @sanjay-fa . We determined that there are two issues:

* what is the behavior of a tenant scoped keymaster updating a different key. This may have changed between 1.26.1 and what is currently on master.
* what should be the behavior of a tenant scoped keymaster updating a different key. This should be that the `tenantId` field is *not* required (or even allowed) since it is silly to require or allow a value which cannot be changed.